# Hackathon

# EDA Processing

### importing libraries

In [1]:
import pandas as pd
import numpy as np

C:\Users\omkar\AppData\Roaming\Python\Python39\site-packages\pandas\core\computation\expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.8.1' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED


### Reading CSV file

In [2]:
df = pd.read_csv("zomato_data.csv")

In [3]:
df.head() # gives first 5 rows

online_order book_table   rate  votes            rest_type  \
0          Yes        Yes  4.1/5    775        Casual Dining   
1          Yes         No  4.1/5    787        Casual Dining   
2          Yes         No  3.8/5    918  Cafe, Casual Dining   
3           No         No  3.7/5     88          Quick Bites   
4           No         No  3.8/5    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                       800        Buffet   
1     Chinese, North Indian, Thai                       800        Buffet   
2          Cafe, Mexican, Italian                       800        Buffet   
3      South Indian, North Indian                       300        Buffet   
4        North Indian, Rajasthani                       600        Buffet   

  listed_incity  
0  Banashankari  
1  Banashankari  
2  Banashankari  
3  Banashankari  
4  Banashankari

In [4]:
# rows and columns of dataset.. There are 51717 rows and 10 columns
df.shape

(51717, 10)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   online_order               51717 non-null  object
 1   book_table                 51717 non-null  object
 2   rate                       43942 non-null  object
 3   votes                      51717 non-null  int64 
 4   rest_type                  51490 non-null  object
 5   dish_liked                 23639 non-null  object
 6   cuisines                   51672 non-null  object
 7   approx_costfor_two_people  51371 non-null  object
 8   listed_intype              51717 non-null  object
 9   listed_incity              51717 non-null  object
dtypes: int64(1), object(9)
memory usage: 3.9+ MB


### Finding null values

In [6]:
df.isna().sum()

online_order                     0
book_table                       0
rate                          7775
votes                            0
rest_type                      227
dish_liked                   28078
cuisines                        45
approx_costfor_two_people      346
listed_intype                    0
listed_incity                    0
dtype: int64

## Data Cleaning & Preprocessing

### Step 1: Rating Column (rate)

In [7]:
df[df['rate'] == '-']

online_order book_table rate  votes      rest_type dish_liked  \
3068            No         No    -      0    Quick Bites        NaN   
3069            No         No    -      0    Quick Bites        NaN   
3373           Yes         No    -      0    Quick Bites        NaN   
3378           Yes         No    -      0    Quick Bites        NaN   
3387            No         No    -      0    Quick Bites        NaN   
...            ...        ...  ...    ...            ...        ...   
33454           No         No    -      0    Quick Bites        NaN   
33463           No         No    -      0  Casual Dining        NaN   
45646          Yes         No    -      0    Quick Bites        NaN   
45647          Yes         No    -      0  Beverage Shop        NaN   
48139           No         No    -      0           Cafe        NaN   

                                             cuisines  \
3068                                     North Indian   
3069              North Indian, Chinese, South Indian   
3373                                            Pizza   
3378                            Chinese, North Indian   
3387                 North Indian, Biryani, Fast Food   
...                                               ...   
33454             South Indian, North Indian, Chinese   
33463     North Indian, South Indian, Chinese, Kerala   
45646  South Indian, North Indian, Chinese, Beverages   
45647                                       Beverages   
48139              North Indian, South American, Cafe   

      approx_costfor_two_people listed_intype          listed_incity  
3068                        400      Delivery           Basavanagudi  
3069                        NaN      Delivery           Basavanagudi  
3373                        300      Dine-out           Basavanagudi  
3378                        300      Dine-out           Basavanagudi  
3387                        300      Dine-out           Basavanagudi  
...                         ...           ...                    ...  
33454                       300      Delivery  Koramangala 6th Block  
33463                       500      Delivery  Koramangala 6th Block  
45646                       400      Dine-out       Old Airport Road  
45647                       250      Dine-out       Old Airport Road  
48139                       200      Dine-out         Residency Road  

[69 rows x 10 columns]

- As there are 69 rows which contain "-" in the rate column.
- Our first task is to replace the "-" with nan value.

In [8]:
df['rate'].replace("-", np.nan, inplace = True)

C:\Users\omkar\AppData\Local\Temp\ipykernel_13508\3081124488.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['rate'].replace("-", np.nan, inplace = True)


In [9]:
df[df['rate'] == '-']

Empty DataFrame
Columns: [online_order, book_table, rate, votes, rest_type, dish_liked, cuisines, approx_costfor_two_people, listed_intype, listed_incity]
Index: []

- As you can see that now there are no "-" values in the rate column.

Next task is to remove /5 from the rate column

In [10]:
df['rate']

0         4.1/5
1         4.1/5
2         3.8/5
3         3.7/5
4         3.8/5
          ...  
51712    3.6 /5
51713       NaN
51714       NaN
51715    4.3 /5
51716    3.4 /5
Name: rate, Length: 51717, dtype: object

- As we can see that there are some spaces in between. Ex: index no. 51712: 3.6 /5 after 3.6 there is space and then there is /5 so first we will remove that white space

In [11]:
df['rate'] = df['rate'].str.replace(' ', '', regex=False)

In [12]:
df['rate']

0        4.1/5
1        4.1/5
2        3.8/5
3        3.7/5
4        3.8/5
         ...  
51712    3.6/5
51713      NaN
51714      NaN
51715    4.3/5
51716    3.4/5
Name: rate, Length: 51717, dtype: object

- So now we can see that whitespace is removed

In [13]:
df['rate'] = df['rate'].str.replace('/5', '', regex=False)

In [14]:
df['rate']

0        4.1
1        4.1
2        3.8
3        3.7
4        3.8
        ... 
51712    3.6
51713    NaN
51714    NaN
51715    4.3
51716    3.4
Name: rate, Length: 51717, dtype: object

- removed '/5' from the column rate

Next is to convert str to numeric

In [16]:
df['rate'] = df['rate'].astype(float)

ValueError: could not convert string to float: 'NEW'

In [17]:
df[df['rate'] == 'NEW']

online_order book_table rate  votes                rest_type dish_liked  \
72              No         No  NEW      0              Quick Bites        NaN   
75              No         No  NEW      0       Takeaway, Delivery        NaN   
110             No         No  NEW      0       Takeaway, Delivery        NaN   
130             No         No  NEW      0       Takeaway, Delivery        NaN   
131             No         No  NEW      0              Quick Bites        NaN   
...            ...        ...  ...    ...                      ...        ...   
51585           No         No  NEW      0              Quick Bites        NaN   
51586          Yes         No  NEW      0              Quick Bites        NaN   
51603          Yes         No  NEW      0  Quick Bites, Food Court        NaN   
51604           No         No  NEW      0       Takeaway, Delivery        NaN   
51605           No         No  NEW      0              Quick Bites        NaN   

                                  cuisines approx_costfor_two_people  \
72                            North Indian                       150   
75                  Street Food, Fast Food                       500   
110             Sandwich, Pizza, Beverages                       200   
130                Biryani, Rolls, Chinese                       200   
131                           North Indian                       100   
...                                    ...                       ...   
51585                         North Indian                       200   
51586                              Chinese                       200   
51603  South Indian, Chinese, North Indian                       250   
51604                     Biryani, Mughlai                       400   
51605                Chinese, North Indian                       300   

      listed_intype listed_incity  
72         Delivery  Banashankari  
75         Delivery  Banashankari  
110        Delivery  Banashankari  
130        Delivery  Banashankari  
131        Delivery  Banashankari  
...             ...           ...  
51585      Dine-out    Whitefield  
51586      Dine-out    Whitefield  
51603      Dine-out    Whitefield  
51604      Dine-out    Whitefield  
51605      Dine-out    Whitefield  

[2208 rows x 10 columns]

In [18]:
df['rate'] = df['rate'].replace('NEW', np.nan)

In [19]:
df[df['rate'] == 'NEW']

Empty DataFrame
Columns: [online_order, book_table, rate, votes, rest_type, dish_liked, cuisines, approx_costfor_two_people, listed_intype, listed_incity]
Index: []

In [20]:
df['rate'] = df['rate'].astype(float)

In [21]:
df['rate'].dtype

dtype('float64')

- converted str to float

Next is to replace NaN value with median of rate...

In [22]:
df['rate'].fillna(df['rate'].median(), inplace = True)

C:\Users\omkar\AppData\Local\Temp\ipykernel_13508\1603575254.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['rate'].fillna(df['rate'].median(), inplace = True)


In [23]:
df['rate'].isna().sum()

0

- As you can see there are no NaN values present now...

### Step 2: Cost Column (approx_costfor_two_people)

First task is to remove ',' from the string

In [24]:
df['approx_costfor_two_people']

0          800
1          800
2          800
3          300
4          600
         ...  
51712    1,500
51713      600
51714    2,000
51715    2,500
51716    1,500
Name: approx_costfor_two_people, Length: 51717, dtype: object

In [25]:
df['approx_costfor_two_people'].str.replace(',', '')

0         800
1         800
2         800
3         300
4         600
         ... 
51712    1500
51713     600
51714    2000
51715    2500
51716    1500
Name: approx_costfor_two_people, Length: 51717, dtype: object

- As you can see the 1,500 was replaced by 1500

In [26]:
df['approx_costfor_two_people'] = df['approx_costfor_two_people'].str.replace(',', '')

In [27]:
df['approx_costfor_two_people']

0         800
1         800
2         800
3         300
4         600
         ... 
51712    1500
51713     600
51714    2000
51715    2500
51716    1500
Name: approx_costfor_two_people, Length: 51717, dtype: object

Second Task is to convert to numeric

In [28]:
df['approx_costfor_two_people'].astype(float)

0         800.0
1         800.0
2         800.0
3         300.0
4         600.0
          ...  
51712    1500.0
51713     600.0
51714    2000.0
51715    2500.0
51716    1500.0
Name: approx_costfor_two_people, Length: 51717, dtype: float64

In [29]:
df['approx_costfor_two_people'] = df['approx_costfor_two_people'].astype(float)

In [30]:
df['approx_costfor_two_people'].dtype

dtype('float64')

Third Task is to fill the missing value with median

In [31]:
df['approx_costfor_two_people'].isna().sum()

346

- There are 346 missing values present so let's replace it with medium value

In [32]:
df['approx_costfor_two_people'].fillna(df['approx_costfor_two_people'].median())

0         800.0
1         800.0
2         800.0
3         300.0
4         600.0
          ...  
51712    1500.0
51713     600.0
51714    2000.0
51715    2500.0
51716    1500.0
Name: approx_costfor_two_people, Length: 51717, dtype: float64

In [33]:
df['approx_costfor_two_people'].fillna(df['approx_costfor_two_people'].median(), inplace = True)

C:\Users\omkar\AppData\Local\Temp\ipykernel_13508\4256619193.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['approx_costfor_two_people'].fillna(df['approx_costfor_two_people'].median(), inplace = True)


In [34]:
df['approx_costfor_two_people'].isna().sum()

0

- Now we replace the NaN values with Median cost value

### Step 3: Categorical Columns 

#### Dish_liked column

In [35]:
df['dish_liked']

0        Pasta, Lunch Buffet, Masala Papad, Paneer Laja...
1        Momos, Lunch Buffet, Chocolate Nirvana, Thai G...
2        Churros, Cannelloni, Minestrone Soup, Hot Choc...
3                                              Masala Dosa
4                                      Panipuri, Gol Gappe
                               ...                        
51712                                                  NaN
51713                                                  NaN
51714                                                  NaN
51715                         Cocktails, Pizza, Buttermilk
51716                                                  NaN
Name: dish_liked, Length: 51717, dtype: object

In [36]:
df['dish_liked'].isna().sum()

28078

Our task is to replace the NaN values with "Not Available"

In [37]:
df['dish_liked'].fillna('Not Available')

0        Pasta, Lunch Buffet, Masala Papad, Paneer Laja...
1        Momos, Lunch Buffet, Chocolate Nirvana, Thai G...
2        Churros, Cannelloni, Minestrone Soup, Hot Choc...
3                                              Masala Dosa
4                                      Panipuri, Gol Gappe
                               ...                        
51712                                        Not Available
51713                                        Not Available
51714                                        Not Available
51715                         Cocktails, Pizza, Buttermilk
51716                                        Not Available
Name: dish_liked, Length: 51717, dtype: object

- As you can see that in the index `51712` `NaN` was present first and now it is replaced with `Not Available`


- Now as it is confired that what we want so let's save it using `inplace = True`

In [38]:
df['dish_liked'].fillna('Not Available', inplace = True)

In [39]:
df['dish_liked'].isna().sum()

0

- Yeah!! Successfully replaced `NaN` values with `Not Applicable`

#### Cuisines column

In [40]:
df['cuisines']

0                North Indian, Mughlai, Chinese
1                   Chinese, North Indian, Thai
2                        Cafe, Mexican, Italian
3                    South Indian, North Indian
4                      North Indian, Rajasthani
                          ...                  
51712                               Continental
51713                               Finger Food
51714                               Finger Food
51715                               Finger Food
51716    Finger Food, North Indian, Continental
Name: cuisines, Length: 51717, dtype: object

In [41]:
df['cuisines'].isna().sum()

45

- There are `45` rows which contain `NaN` values.


- Our task is to replace those `45` `NaN` values with `Other`

In [42]:
df['cuisines'].fillna('Other', inplace = True)

In [43]:
df['cuisines'].isna().sum()

0

- Yeah!! Successfully replaced 45 `NaN` value with `Other`

#### rest_type column

In [44]:
df['rest_type']

0              Casual Dining
1              Casual Dining
2        Cafe, Casual Dining
3                Quick Bites
4              Casual Dining
                ...         
51712                    Bar
51713                    Bar
51714                    Bar
51715                    Bar
51716     Bar, Casual Dining
Name: rest_type, Length: 51717, dtype: object

In [45]:
df['rest_type'].isna().sum()

227

- There are `227` rows which contain `NaN` values.


- Our task is to replace those `227` `NaN` values with `Unknown`

In [46]:
df['rest_type'].fillna('Unknown', inplace = True)

In [47]:
df['rest_type'].isna().sum()

0

- Yeah!! Successfully replaced 227 `NaN` value with `Unknown`

### Step 4: Votes Column

Our task is to fill the missing values with median

In [48]:
df['votes'].isna().sum()

0

In [49]:
df.isna().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

- As we can see that there are no missing values in any of the columns.


- So we will not replace the 'NaN' value of vote column with median

If the `NaN` values were present then the code you need to use will be:

```
df['votes'].fillna(df['votes'].median(), inplace = True)
```

### Step 5: Binary Encoding

#### online_order column

In [50]:
df['online_order']

0        Yes
1        Yes
2        Yes
3         No
4         No
        ... 
51712     No
51713     No
51714     No
51715     No
51716     No
Name: online_order, Length: 51717, dtype: object

- Our task is to convert `Yes` to `1` and `No` to `0`

In [51]:
df['online_order'].map({'Yes': 1, 'No': 0})

0        1
1        1
2        1
3        0
4        0
        ..
51712    0
51713    0
51714    0
51715    0
51716    0
Name: online_order, Length: 51717, dtype: int64

- As we can see that `Yes` is converted to `1` and `No` is converted to `0` 


- So let's save this

In [52]:
df['online_order'] = df['online_order'].map({'Yes': 1, 'No': 0})

In [53]:
df['online_order']

0        1
1        1
2        1
3        0
4        0
        ..
51712    0
51713    0
51714    0
51715    0
51716    0
Name: online_order, Length: 51717, dtype: int64

- Yeah!! Successfully converted `Yes` to `1` adn `No` to `0`

#### book_table column

In [54]:
df['book_table']

0        Yes
1         No
2         No
3         No
4         No
        ... 
51712     No
51713     No
51714     No
51715    Yes
51716     No
Name: book_table, Length: 51717, dtype: object

- Our task is to convert `Yes` to `1` and `No` to `0`

In [55]:
df['book_table'].map({'Yes': 1, 'No': 0})

0        1
1        0
2        0
3        0
4        0
        ..
51712    0
51713    0
51714    0
51715    1
51716    0
Name: book_table, Length: 51717, dtype: int64

- As we can see that `Yes` is converted to `1` and `No` is converted to `0` 


- So let's save this

In [56]:
df['book_table'] = df['book_table'].map({'Yes': 1, 'No': 0})

In [57]:
df['book_table']

0        1
1        0
2        0
3        0
4        0
        ..
51712    0
51713    0
51714    0
51715    1
51716    0
Name: book_table, Length: 51717, dtype: int64

### Step 6: Data Type Conversion

Our task is to ensure the following conversions:

- rate → float


- votes → integer


- approx_costfor_two_people → integer

In [58]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  int64  
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(3), object(5)
memory usage: 3.9+ MB


- As we can see in the above info, `rate` is having `float` datatype, and `votes` is having `int` datatype which will satisfy our condition.


- But the `approx_costfor_two_people` column datatype is `float`. So now we need to convert the `float` datatype to `int` to satisfy our conditions

In [59]:
df['approx_costfor_two_people'] = df['approx_costfor_two_people'].astype(int)

In [60]:
df['approx_costfor_two_people'].dtype

dtype('int32')

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  int64  
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  int32  
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(1), int32(1), int64(3), object(5)
memory usage: 3.7+ MB


- Yeah!! Successfully converted `float` to `int`

In [62]:
df.isna().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

- There are no missing values in the dataset

### Saving the cleaned dataset into csv format

In [63]:
df.to_csv('cleaned_zomato_details.csv', index = None)

## Merging Datasets

In [64]:
df1 = pd.read_csv('cleaned_zomato_details.csv')

In [65]:
df1.head()

online_order  book_table  rate  votes            rest_type  \
0             1           1   4.1    775        Casual Dining   
1             1           0   4.1    787        Casual Dining   
2             1           0   3.8    918  Cafe, Casual Dining   
3             0           0   3.7     88          Quick Bites   
4             0           0   3.8    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                        800        Buffet   
1     Chinese, North Indian, Thai                        800        Buffet   
2          Cafe, Mexican, Italian                        800        Buffet   
3      South Indian, North Indian                        300        Buffet   
4        North Indian, Rajasthani                        600        Buffet   

  listed_incity  
0  Banashankari  
1  Banashankari  
2  Banashankari  
3  Banashankari  
4  Banashankari

In [66]:
df1.isna().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [75]:
df1.shape

(51717, 10)

In [67]:
df2 = pd.read_csv('Geographical Coordinates.csv')

In [68]:
df2.head()

listed_incity   Latitude  Longitude
0       Banashankari  12.939333  77.553982
1  Bannerghatta Road  12.952660  77.605048
2       Basavanagudi  12.941726  77.575502
3          Bellandur  12.925352  77.675941
4       Brigade Road  12.967358  77.606435

In [70]:
df2.isna().sum()

listed_incity    0
Latitude         0
Longitude        0
dtype: int64

In [76]:
df2.shape

(26, 3)

In [71]:
merged_df = pd.merge(df1, df2, on='listed_incity', how='left')

In [72]:
merged_df.head()

online_order  book_table  rate  votes            rest_type  \
0             1           1   4.1    775        Casual Dining   
1             1           0   4.1    787        Casual Dining   
2             1           0   3.8    918  Cafe, Casual Dining   
3             0           0   3.7     88          Quick Bites   
4             0           0   3.8    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                        800        Buffet   
1     Chinese, North Indian, Thai                        800        Buffet   
2          Cafe, Mexican, Italian                        800        Buffet   
3      South Indian, North Indian                        300        Buffet   
4        North Indian, Rajasthani                        600        Buffet   

  listed_incity   Latitude  Longitude  
0  Banashankari  12.939333  77.553982  
1  Banashankari  12.939333  77.553982  
2  Banashankari  12.939333  77.553982  
3  Banashankari  12.939333  77.553982  
4  Banashankari  12.939333  77.553982

In [73]:
merged_df.isna().sum()

online_order                    0
book_table                      0
rate                            0
votes                           0
rest_type                       0
dish_liked                      0
cuisines                        0
approx_costfor_two_people       0
listed_intype                   0
listed_incity                   0
Latitude                     5580
Longitude                    5580
dtype: int64

### Saving merged dataset into csv file

In [77]:
merged_df.to_csv('merged_data.csv', index = False)

In [78]:
df2 = pd.read_csv('merged_data.csv')

In [79]:
df2.head()

online_order  book_table  rate  votes            rest_type  \
0             1           1   4.1    775        Casual Dining   
1             1           0   4.1    787        Casual Dining   
2             1           0   3.8    918  Cafe, Casual Dining   
3             0           0   3.7     88          Quick Bites   
4             0           0   3.8    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                        800        Buffet   
1     Chinese, North Indian, Thai                        800        Buffet   
2          Cafe, Mexican, Italian                        800        Buffet   
3      South Indian, North Indian                        300        Buffet   
4        North Indian, Rajasthani                        600        Buffet   

  listed_incity   Latitude  Longitude  
0  Banashankari  12.939333  77.553982  
1  Banashankari  12.939333  77.553982  
2  Banashankari  12.939333  77.553982  
3  Banashankari  12.939333  77.553982  
4  Banashankari  12.939333  77.553982

In [80]:
df2.isna().sum()

online_order                    0
book_table                      0
rate                            0
votes                           0
rest_type                       0
dish_liked                      0
cuisines                        0
approx_costfor_two_people       0
listed_intype                   0
listed_incity                   0
Latitude                     5580
Longitude                    5580
dtype: int64

## Cuisine-Specific Map (Italian Restaurants)

### Installing folium

In [ ]:
!pip install folium

### Importing Libraries

In [81]:
import folium
from folium.plugins import MarkerCluster
from IPython.display import IFrame

### reading data

In [82]:
df_merged = pd.read_csv('merged_data.csv')

In [83]:
df_merged.head()

online_order  book_table  rate  votes            rest_type  \
0             1           1   4.1    775        Casual Dining   
1             1           0   4.1    787        Casual Dining   
2             1           0   3.8    918  Cafe, Casual Dining   
3             0           0   3.7     88          Quick Bites   
4             0           0   3.8    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                        800        Buffet   
1     Chinese, North Indian, Thai                        800        Buffet   
2          Cafe, Mexican, Italian                        800        Buffet   
3      South Indian, North Indian                        300        Buffet   
4        North Indian, Rajasthani                        600        Buffet   

  listed_incity   Latitude  Longitude  
0  Banashankari  12.939333  77.553982  
1  Banashankari  12.939333  77.553982  
2  Banashankari  12.939333  77.553982  
3  Banashankari  12.939333  77.553982  
4  Banashankari  12.939333  77.553982

### Cuisine - Specific Map -- Italian Restaurants

In [84]:
df_merged.isna().sum()

online_order                    0
book_table                      0
rate                            0
votes                           0
rest_type                       0
dish_liked                      0
cuisines                        0
approx_costfor_two_people       0
listed_intype                   0
listed_incity                   0
Latitude                     5580
Longitude                    5580
dtype: int64

#### Filter for Italian Restaurants (case-insensitive)

In [85]:
italian_df = df_merged[df_merged['cuisines'].str.contains('Italian', case=False, na=False)]

In [86]:
italian_df.head()

online_order  book_table  rate  votes            rest_type  \
2              1           0   3.8    918  Cafe, Casual Dining   
7              1           1   4.6   2556  Casual Dining, Cafe   
8              1           0   4.0    324                 Cafe   
9              1           0   4.2    504                 Cafe   
11             1           1   4.2    150                 Cafe   

                                           dish_liked  \
2   Churros, Cannelloni, Minestrone Soup, Hot Choc...   
7   Farmhouse Pizza, Chocolate Banana, Virgin Moji...   
8   Pizza, Mocktails, Coffee, Nachos, Salad, Pasta...   
9   Waffles, Pasta, Coleslaw Sandwich, Choco Waffl...   
11  Mocktails, Peri Fries, Lasagne, Pizza, Chicken...   

                                    cuisines  approx_costfor_two_people  \
2                     Cafe, Mexican, Italian                        800   
7                       Pizza, Cafe, Italian                        600   
8                 Cafe, Italian, Continental                        700   
9   Cafe, Mexican, Italian, Momos, Beverages                        550   
11                Cafe, Italian, Continental                        600   

   listed_intype listed_incity   Latitude  Longitude  
2         Buffet  Banashankari  12.939333  77.553982  
7          Cafes  Banashankari  12.939333  77.553982  
8          Cafes  Banashankari  12.939333  77.553982  
9          Cafes  Banashankari  12.939333  77.553982  
11         Cafes  Banashankari  12.939333  77.553982

In [87]:
italian_df.isna().sum()

online_order                   0
book_table                     0
rate                           0
votes                          0
rest_type                      0
dish_liked                     0
cuisines                       0
approx_costfor_two_people      0
listed_intype                  0
listed_incity                  0
Latitude                     343
Longitude                    343
dtype: int64

In [88]:
italian_df.shape

(3389, 12)

#### Drop rows with missing latitude or longitude

In [89]:
italian_df = italian_df.dropna(subset=['Latitude', 'Longitude'])

In [90]:
italian_df.isna().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
Latitude                     0
Longitude                    0
dtype: int64

#### Initialize a map centered on the average location of Italian restaurants

In [91]:
center_lat = italian_df['Latitude'].mean()
center_lon = italian_df['Longitude'].mean()
italian_map = folium.Map(location=[center_lat, center_lon], zoom_start=12)

#### Add purple markers for each Italian restaurant

In [102]:
for _, row in italian_df.iterrows():
    folium.Marker(
        location=[row['Latitude'], row['Longitude']],
        popup=f"City: {row['listed_incity']} <br> Cuisine: {row['cuisines']} <br> Ratings: {row['rate']}",
        icon=folium.Icon(color='purple')
    ).add_to(italian_map)

#### Save the map as an HTML file

In [103]:
italian_map.save('italian_restaurants_map.html')

### Display using IPython

In [248]:
IFrame('italian_restaurants_map.html', width=800, height=600)

## Restaurant Density Map

#### Initialize the map centered on Bangalore

In [105]:
bangalore_center = [12.9716, 77.5946]
restaurant_map = folium.Map(location=bangalore_center, zoom_start=12)

#### Create a MarkerCluster

In [106]:
marker_cluster = MarkerCluster().add_to(restaurant_map)

#### Add restaurant markers to the map

In [107]:
df_valid = df_merged.dropna(subset=['Latitude', 'Longitude'])

In [108]:
for _, row in df_valid.iterrows():
    folium.Marker(
        location=[row['Latitude'], row['Longitude']],
        popup=f"<b>City:</b> {row['listed_incity']}<br> <b>Rating:</b> {row['rate']}<br> <b>Cost for Two:</b> ₹{row['approx_costfor_two_people']}"
    ).add_to(marker_cluster)

#### Save the map as an HTML file

In [109]:
restaurant_map.save('restaurant_density_map.html')

In [249]:
IFrame('restaurant_density_map.html', width=800, height=600)

## MCQs

In [144]:
mcq_df = pd.read_csv('cleaned_zomato_details.csv')

In [145]:
mcq_df.head()

online_order  book_table  rate  votes            rest_type  \
0             1           1   4.1    775        Casual Dining   
1             1           0   4.1    787        Casual Dining   
2             1           0   3.8    918  Cafe, Casual Dining   
3             0           0   3.7     88          Quick Bites   
4             0           0   3.8    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                        800        Buffet   
1     Chinese, North Indian, Thai                        800        Buffet   
2          Cafe, Mexican, Italian                        800        Buffet   
3      South Indian, North Indian                        300        Buffet   
4        North Indian, Rajasthani                        600        Buffet   

  listed_incity  
0  Banashankari  
1  Banashankari  
2  Banashankari  
3  Banashankari  
4  Banashankari

### Q1: What is the shape of the given dataset?

In [146]:
mcq_df.shape

(51717, 10)

### Q2: How many restaurants serve North Indian cuisine?

In [147]:
north_indian_restaurants = mcq_df[mcq_df['cuisines'].str.contains('North Indian', na=False)]

In [148]:
north_indian_restaurants.shape[0]

21085

### Q3: What cuisine is most commonly offered by restaurants in Bangalore?

In [149]:
set(mcq_df['listed_incity'])

{'BTM',
 'Banashankari',
 'Bannerghatta Road',
 'Basavanagudi',
 'Bellandur',
 'Brigade Road',
 'Brookefield',
 'Church Street',
 'Electronic City',
 'Frazer Town',
 'HSR',
 'Indiranagar',
 'JP Nagar',
 'Jayanagar',
 'Kalyan Nagar',
 'Kammanahalli',
 'Koramangala 4th Block',
 'Koramangala 5th Block',
 'Koramangala 6th Block',
 'Koramangala 7th Block',
 'Lavelle Road',
 'MG Road',
 'Malleshwaram',
 'Marathahalli',
 'New BEL Road',
 'Old Airport Road',
 'Rajajinagar',
 'Residency Road',
 'Sarjapur Road',
 'Whitefield'}

In [150]:
mcq_df['cuisines'] = mcq_df['cuisines'].str.strip().str.split(', ')

In [151]:
mcq_df['cuisines']

0                [North Indian, Mughlai, Chinese]
1                   [Chinese, North Indian, Thai]
2                        [Cafe, Mexican, Italian]
3                    [South Indian, North Indian]
4                      [North Indian, Rajasthani]
                           ...                   
51712                               [Continental]
51713                               [Finger Food]
51714                               [Finger Food]
51715                               [Finger Food]
51716    [Finger Food, North Indian, Continental]
Name: cuisines, Length: 51717, dtype: object

In [153]:
df_exploded = mcq_df.explode('cuisines')

In [154]:
df_exploded

online_order  book_table  rate  votes           rest_type  \
0                 1           1   4.1    775       Casual Dining   
0                 1           1   4.1    775       Casual Dining   
0                 1           1   4.1    775       Casual Dining   
1                 1           0   4.1    787       Casual Dining   
1                 1           0   4.1    787       Casual Dining   
...             ...         ...   ...    ...                 ...   
51714             0           0   3.7      0                 Bar   
51715             0           1   4.3    236                 Bar   
51716             0           0   3.4     13  Bar, Casual Dining   
51716             0           0   3.4     13  Bar, Casual Dining   
51716             0           0   3.4     13  Bar, Casual Dining   

                                              dish_liked      cuisines  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...  North Indian   
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...       Mughlai   
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...       Chinese   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...       Chinese   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...  North Indian   
...                                                  ...           ...   
51714                                      Not Available   Finger Food   
51715                       Cocktails, Pizza, Buttermilk   Finger Food   
51716                                      Not Available   Finger Food   
51716                                      Not Available  North Indian   
51716                                      Not Available   Continental   

       approx_costfor_two_people  listed_intype listed_incity  
0                            800         Buffet  Banashankari  
0                            800         Buffet  Banashankari  
0                            800         Buffet  Banashankari  
1                            800         Buffet  Banashankari  
1                            800         Buffet  Banashankari  
...                          ...            ...           ...  
51714                       2000  Pubs and bars    Whitefield  
51715                       2500  Pubs and bars    Whitefield  
51716                       1500  Pubs and bars    Whitefield  
51716                       1500  Pubs and bars    Whitefield  
51716                       1500  Pubs and bars    Whitefield  

[126864 rows x 10 columns]

In [155]:
df_exploded = df_exploded[df_exploded['cuisines'].str.strip() != '']

In [156]:
most_common = df_exploded['cuisines'].value_counts().idxmax()
count = df_exploded['cuisines'].value_counts().max()

In [158]:
print(f"Most common cuisine in Bangalore: {most_common} ({count} restaurants)")

Most common cuisine in Bangalore: North Indian (21085 restaurants)


### Q4: Which locality in Bangalore has the highest average cost for dining (for two people)?

In [159]:
avg_cost_by_locality = (
    mcq_df.groupby('listed_incity')['approx_costfor_two_people']
    .mean()
    .sort_values(ascending=False)
)

In [160]:
highest_locality = avg_cost_by_locality.idxmax()
highest_cost = avg_cost_by_locality.max()

In [161]:
print(f"Locality with the highest average cost for two: {highest_locality} (₹{highest_cost:.2f})")

Locality with the highest average cost for two: Church Street (₹770.36)


### Q5: Which restaurant type has the top rating with over 1000 votes?

In [162]:
df_filtered = df[df['votes'] > 1000]

In [163]:
avg_rating_by_type = (
    df_filtered.groupby('rest_type')['rate']
    .mean()
    .sort_values(ascending=False)
)

In [164]:
top_type = avg_rating_by_type.idxmax()
top_rating = avg_rating_by_type.max()


In [165]:
print(f"Top rated restaurant type with over 1000 votes: {top_type} (Avg rating: {top_rating:.2f})")

Top rated restaurant type with over 1000 votes: Bakery (Avg rating: 4.80)


### Q6: How much does it cost at minimum to eat out in Bangalore?


In [167]:
min_cost = mcq_df['approx_costfor_two_people'].min()

In [168]:
print(f"Minimum cost for dining out in Bangalore: ₹{min_cost:.2f}")

Minimum cost for dining out in Bangalore: ₹40.00


### Q7: What percentage of total online orders is received by restaurants in Banashankari?

In [169]:
banashankari_df = mcq_df[mcq_df['listed_incity'].str.lower() == 'banashankari']

In [170]:
banashankari_df

online_order  book_table  rate  votes            rest_type  \
0               1           1   4.1    775        Casual Dining   
1               1           0   4.1    787        Casual Dining   
2               1           0   3.8    918  Cafe, Casual Dining   
3               0           0   3.7     88          Quick Bites   
4               0           0   3.8    166        Casual Dining   
..            ...         ...   ...    ...                  ...   
858             0           0   3.3     62                  Bar   
859             0           0   3.1     30                  Bar   
860             0           0   3.7      0                  Bar   
861             0           0   3.0     98                  Bar   
862             0           0   2.7    377                  Pub   

                                            dish_liked  \
0    Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1    Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2    Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                          Masala Dosa   
4                                  Panipuri, Gol Gappe   
..                                                 ...   
858                                     Chilli Chicken   
859                                      Not Available   
860                                      Not Available   
861            Paneer Tikka, Dal Kichadi, French Fries   
862  Thai Chicken, Chicken Lollipop, Cocktails, Jee...   

                                  cuisines  approx_costfor_two_people  \
0         [North Indian, Mughlai, Chinese]                        800   
1            [Chinese, North Indian, Thai]                        800   
2                 [Cafe, Mexican, Italian]                        800   
3             [South Indian, North Indian]                        300   
4               [North Indian, Rajasthani]                        600   
..                                     ...                        ...   
858  [South Indian, North Indian, Chinese]                        800   
859                [Chinese, North Indian]                       1000   
860               [Fast Food, Finger Food]                        500   
861                [North Indian, Chinese]                        800   
862          [North Indian, Chinese, Thai]                       1100   

          listed_intype listed_incity  
0                Buffet  Banashankari  
1                Buffet  Banashankari  
2                Buffet  Banashankari  
3                Buffet  Banashankari  
4                Buffet  Banashankari  
..                  ...           ...  
858  Drinks & nightlife  Banashankari  
859  Drinks & nightlife  Banashankari  
860  Drinks & nightlife  Banashankari  
861  Drinks & nightlife  Banashankari  
862  Drinks & nightlife  Banashankari  

[863 rows x 10 columns]

In [179]:
online_count = banashankari_df[banashankari_df['online_order'] == 1].shape[0]

In [180]:
online_count

546

In [244]:
total_count = banashankari_df.shape[0]

In [245]:
total_count

863

In [246]:
percentage = (online_count/total_count)*100

In [247]:
percentage

63.26767091541136

### Q8: Which locality has the most restaurants with over 500 votes and a rating below 3.0?

In [191]:
filtered_df = mcq_df[(mcq_df['votes'] > 500) & (mcq_df['rate'] < 3.0)]

In [192]:
filtered_df.shape

(43, 10)

In [193]:
locality_counts = filtered_df['listed_incity'].value_counts()

In [194]:
top_locality = locality_counts.idxmax()
top_count = locality_counts.max()

In [195]:
print(f"Locality with most restaurants (votes > 500 & rating < 3.0): {top_locality} ({top_count} restaurants)")

Locality with most restaurants (votes > 500 & rating < 3.0): Brookefield (8 restaurants)


### Q9: Which locality in Bangalore should Zomato target for expansion based on restaurant type diversity?

In [196]:
mcq_df['listed_incity'] = mcq_df['listed_incity'].str.strip()
mcq_df['rest_type'] = mcq_df['rest_type'].str.strip()

In [197]:
mcq_df

online_order  book_table  rate  votes            rest_type  \
0                 1           1   4.1    775        Casual Dining   
1                 1           0   4.1    787        Casual Dining   
2                 1           0   3.8    918  Cafe, Casual Dining   
3                 0           0   3.7     88          Quick Bites   
4                 0           0   3.8    166        Casual Dining   
...             ...         ...   ...    ...                  ...   
51712             0           0   3.6     27                  Bar   
51713             0           0   3.7      0                  Bar   
51714             0           0   3.7      0                  Bar   
51715             0           1   4.3    236                  Bar   
51716             0           0   3.4     13   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                            Masala Dosa   
4                                    Panipuri, Gol Gappe   
...                                                  ...   
51712                                      Not Available   
51713                                      Not Available   
51714                                      Not Available   
51715                       Cocktails, Pizza, Buttermilk   
51716                                      Not Available   

                                       cuisines  approx_costfor_two_people  \
0              [North Indian, Mughlai, Chinese]                        800   
1                 [Chinese, North Indian, Thai]                        800   
2                      [Cafe, Mexican, Italian]                        800   
3                  [South Indian, North Indian]                        300   
4                    [North Indian, Rajasthani]                        600   
...                                         ...                        ...   
51712                             [Continental]                       1500   
51713                             [Finger Food]                        600   
51714                             [Finger Food]                       2000   
51715                             [Finger Food]                       2500   
51716  [Finger Food, North Indian, Continental]                       1500   

       listed_intype listed_incity  
0             Buffet  Banashankari  
1             Buffet  Banashankari  
2             Buffet  Banashankari  
3             Buffet  Banashankari  
4             Buffet  Banashankari  
...              ...           ...  
51712  Pubs and bars    Whitefield  
51713  Pubs and bars    Whitefield  
51714  Pubs and bars    Whitefield  
51715  Pubs and bars    Whitefield  
51716  Pubs and bars    Whitefield  

[51717 rows x 10 columns]

In [198]:
diversity_by_locality = mcq_df.groupby('listed_incity')['rest_type'].nunique()

In [199]:
diversity_by_locality

listed_incity
BTM                      62
Banashankari             37
Bannerghatta Road        47
Basavanagudi             40
Bellandur                49
Brigade Road             61
Brookefield              50
Church Street            61
Electronic City          35
Frazer Town              51
HSR                      50
Indiranagar              56
JP Nagar                 49
Jayanagar                53
Kalyan Nagar             42
Kammanahalli             42
Koramangala 4th Block    59
Koramangala 5th Block    58
Koramangala 6th Block    59
Koramangala 7th Block    58
Lavelle Road             59
MG Road                  59
Malleshwaram             47
Marathahalli             44
New BEL Road             35
Old Airport Road         50
Rajajinagar              46
Residency Road           57
Sarjapur Road            45
Whitefield               54
Name: rest_type, dtype: int64

In [200]:
# Get the locality with the highest diversity
top_diverse_locality = diversity_by_locality.idxmax()
diverse_count = diversity_by_locality.max()

In [201]:
print(f"Locality with highest restaurant type diversity: {top_diverse_locality} ({diverse_count} types)")

Locality with highest restaurant type diversity: BTM (62 types)


### Q10: What's the average cost difference between buffet and delivery restaurants?

In [203]:
mcq_df['rest_type'] = mcq_df['rest_type'].str.strip()

In [204]:
buffet_df = mcq_df[mcq_df['rest_type'].str.contains('Buffet', case=False)]
delivery_df = mcq_df[mcq_df['rest_type'].str.contains('Delivery', case=False)]

In [205]:
buffet_avg = buffet_df['approx_costfor_two_people'].mean()
delivery_avg = delivery_df['approx_costfor_two_people'].mean()

In [206]:
cost_diff = buffet_avg - delivery_avg

In [207]:
print(f"Avg Buffet Cost: ₹{buffet_avg:.2f}")
print(f"Avg Delivery Cost: ₹{delivery_avg:.2f}")
print(f"Average cost difference (Buffet - Delivery): ₹{cost_diff:.2f}")

Avg Buffet Cost: ₹nan
Avg Delivery Cost: ₹414.92
Average cost difference (Buffet - Delivery): ₹nan


### Q11: What is the maximum number of votes received by any restaurant with online ordering?

In [208]:
online_df = mcq_df[mcq_df['online_order'] == 1]

In [209]:
max_votes = online_df['votes'].max()

In [210]:
print(f"Maximum votes received by any restaurant with online ordering: {int(max_votes)}")

Maximum votes received by any restaurant with online ordering: 16832


### Q12: What is the average rating of restaurants that serve both North Indian and Chinese cuisines?

In [211]:
restaurant_cuisines = df_exploded.groupby(df_exploded.index)['cuisines'].apply(set)

In [212]:
restaurant_cuisines

0                {Chinese, Mughlai, North Indian}
1                   {Thai, North Indian, Chinese}
2                        {Cafe, Mexican, Italian}
3                    {South Indian, North Indian}
4                      {Rajasthani, North Indian}
                           ...                   
51712                               {Continental}
51713                               {Finger Food}
51714                               {Finger Food}
51715                               {Finger Food}
51716    {Finger Food, Continental, North Indian}
Name: cuisines, Length: 51717, dtype: object

In [213]:
both_cuisines_mask = restaurant_cuisines.apply(lambda x: {'North Indian', 'Chinese'}.issubset(x))

In [214]:
both_cuisines_mask

0         True
1         True
2        False
3        False
4        False
         ...  
51712    False
51713    False
51714    False
51715    False
51716    False
Name: cuisines, Length: 51717, dtype: bool

In [215]:
restaurants_with_both = both_cuisines_mask[both_cuisines_mask].index

In [216]:
restaurants_with_both

Index([    0,     1,     6,    29,    31,    36,    38,    46,    47,    49,
       ...
       51662, 51665, 51670, 51673, 51677, 51694, 51701, 51704, 51708, 51711],
      dtype='int64', length=11394)

In [217]:
both_df = df.loc[restaurants_with_both]

In [218]:
both_df

online_order  book_table  rate  votes            rest_type  \
0                 1           1   4.1    775        Casual Dining   
1                 1           0   4.1    787        Casual Dining   
6                 0           0   3.6      8        Casual Dining   
29                0           0   3.8    148  Casual Dining, Cafe   
31                1           0   2.8    506          Quick Bites   
...             ...         ...   ...    ...                  ...   
51694             1           0   4.1   2773                  Pub   
51701             1           1   4.1    673                  Pub   
51704             0           1   4.0    189   Casual Dining, Pub   
51708             0           0   2.8    161   Casual Dining, Bar   
51711             0           0   2.5     81   Casual Dining, Bar   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
6                                          Not Available   
29     Paratha, Sandwich, Chilli Idli, Noodles, Sandw...   
31     Burgers, Lassi, Chicken Grill, Naan, Momos, Ch...   
...                                                  ...   
51694  Cocktails, Chicken Kebab, Murgh Ghee Roast, Vi...   
51701  Cocktails, Wheat Beer, Bruschettas, Pasta, Bur...   
51704                                        Pizza, Beer   
51708  Salads, Coffee, Breakfast Buffet, Halwa, Chick...   
51711                               Biryani, Andhra Meal   

                                                cuisines  \
0                         North Indian, Mughlai, Chinese   
1                            Chinese, North Indian, Thai   
6            North Indian, South Indian, Andhra, Chinese   
29                North Indian, Cafe, Chinese, Fast Food   
31              North Indian, Fast Food, Chinese, Burger   
...                                                  ...   
51694  American, North Indian, Chinese, Finger Food, ...   
51701        Finger Food, North Indian, Chinese, Italian   
51704                       North Indian, Pizza, Chinese   
51708                 Chinese, Continental, North Indian   
51711        Andhra, South Indian, Chinese, North Indian   

       approx_costfor_two_people  listed_intype listed_incity  
0                            800         Buffet  Banashankari  
1                            800         Buffet  Banashankari  
6                            800         Buffet  Banashankari  
29                           550          Cafes  Banashankari  
31                           500       Delivery  Banashankari  
...                          ...            ...           ...  
51694                       1500  Pubs and bars    Whitefield  
51701                       1400  Pubs and bars    Whitefield  
51704                       1500  Pubs and bars    Whitefield  
51708                       1200  Pubs and bars    Whitefield  
51711                        800  Pubs and bars    Whitefield  

[11394 rows x 10 columns]

In [219]:
avg_rating = both_df['rate'].mean()

In [220]:
print(f"Average rating of restaurants serving both North Indian and Chinese cuisines: {avg_rating:.2f}")

Average rating of restaurants serving both North Indian and Chinese cuisines: 3.59


### Q13: What is the most profitable area for Zomato based on potential revenue estimation?

In [223]:
df_online = mcq_df[mcq_df['online_order'] == 1]

In [224]:
area_stats = df_online.groupby('listed_incity').agg({
    'rate': 'mean',                     # Average rating
    'votes': 'sum',                     # Total votes (popularity)
    'approx_costfor_two_people': 'mean', # Average cost per order
    'listed_incity': 'count'            # Number of restaurants
}).rename(columns={'listed_incity': 'restaurant_count'})

In [225]:
area_stats['normalized_rate'] = area_stats['rate'] / area_stats['rate'].max()
area_stats['normalized_votes'] = area_stats['votes'] / area_stats['votes'].max()
area_stats['normalized_cost'] = area_stats['approx_costfor_two_people'] / area_stats['approx_costfor_two_people'].max()

In [226]:
area_stats['revenue_score'] = (
    (1/3) * area_stats['normalized_rate'] +
    (1/3) * area_stats['normalized_votes'] +
    (1/3) * area_stats['normalized_cost']
)

In [227]:
df_online['has_north_indian'] = df_online['cuisines'].str.contains('North Indian', case=False, na=False)
df_online['has_chinese'] = df_online['cuisines'].str.contains('Chinese', case=False, na=False)

C:\Users\omkar\AppData\Local\Temp\ipykernel_13508\568744266.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_online['has_north_indian'] = df_online['cuisines'].str.contains('North Indian', case=False, na=False)
C:\Users\omkar\AppData\Local\Temp\ipykernel_13508\568744266.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_online['has_chinese'] = df_online['cuisines'].str.contains('Chinese', case=False, na=False)


In [228]:
cuisine_stats = df_online.groupby('listed_incity').agg({
    'has_north_indian': 'sum',
    'has_chinese': 'sum'
}).rename(columns={
    'has_north_indian': 'north_indian_count',
    'has_chinese': 'chinese_count'
})

In [229]:
area_stats = area_stats.merge(cuisine_stats, left_index=True, right_index=True, how='left')

# Calculate cuisine boost (proportion of restaurants with popular cuisines)
area_stats['cuisine_boost'] = (area_stats['north_indian_count'] + area_stats['chinese_count']) / area_stats['restaurant_count']
# Adjust revenue score with cuisine boost (10% boost per cuisine proportion)
area_stats['adjusted_revenue_score'] = area_stats['revenue_score'] * (1 + 0.1 * area_stats['cuisine_boost'])

# Step 6: Identify the most profitable area
most_profitable_area = area_stats.sort_values('adjusted_revenue_score', ascending=False).head(1)

# Display results
print("Most Profitable Area for Zomato Based on Revenue Estimation:")
print(most_profitable_area[['rate', 'votes', 'approx_costfor_two_people', 'restaurant_count', 
                           'north_indian_count', 'chinese_count', 'adjusted_revenue_score']])

Most Profitable Area for Zomato Based on Revenue Estimation:
                           rate   votes  approx_costfor_two_people  \
listed_incity                                                        
Koramangala 7th Block  3.769828  614571                 509.977815   

                       restaurant_count  north_indian_count  chinese_count  \
listed_incity                                                                
Koramangala 7th Block              1803                   0              0   

                       adjusted_revenue_score  
listed_incity                                  
Koramangala 7th Block                0.904285  


### Q14: If Zomato wants to reduce customer complaints, which restaurant type should they focus on?

In [230]:
df_online = df[df['online_order'] == 1]

In [231]:
rest_type_stats = df_online.groupby('rest_type').agg({
    'rate': 'mean',          # Average rating
    'votes': 'sum',          # Total votes (popularity)
    'rest_type': 'count'     # Number of restaurants
}).rename(columns={'rest_type': 'restaurant_count'})

In [232]:
rest_type_stats = rest_type_stats[(rest_type_stats['restaurant_count'] >= 10) & (rest_type_stats['votes'] >= 100)]

In [233]:
problematic_rest_type = rest_type_stats.sort_values('rate', ascending=True).head(1)

In [234]:
print("Restaurant Type to Focus on for Reducing Customer Complaints:")
print(problematic_rest_type[['rate', 'votes', 'restaurant_count']])

Restaurant Type to Focus on for Reducing Customer Complaints:
                         rate  votes  restaurant_count
rest_type                                             
Quick Bites, Food Court   3.3   2228                16


### Q15: In which area should Zomato invest by considering high rating (rate > 4.2), high number of votes (> 500) and including online orders?

In [236]:
df_filtered = df[(df['online_order'] == 1) & (df['rate'] > 4.2) & (df['votes'] > 500)]

In [237]:
area_stats = df_filtered.groupby('listed_incity').agg({
    'listed_incity': 'count',  # Count of qualifying restaurants
    'rate': 'mean',            # Average rating
    'votes': 'sum',            # Total votes
    'approx_costfor_two_people': 'mean'  # Average cost
}).rename(columns={'listed_incity': 'qualifying_restaurants'})

In [238]:
top_area = area_stats.sort_values('qualifying_restaurants', ascending=False).head(1)

In [239]:
print("Best Area for Zomato Investment (High Ratings, High Votes, Online Orders):")
print(top_area[['qualifying_restaurants', 'rate', 'votes', 'approx_costfor_two_people']])

Best Area for Zomato Investment (High Ratings, High Votes, Online Orders):
               qualifying_restaurants      rate   votes  \
listed_incity                                             
MG Road                            97  4.402062  181584   

               approx_costfor_two_people  
listed_incity                             
MG Road                      1268.041237  
